File: 2019/C++/src/intcode.cpp
/**
 * \file intcode.cpp
 * \brief AoC 2019 (C++ version) - Intcode interpreter
 * \author Mina Pêcheux
 * \date 2019
 *
 * Intcode interpreter used multiple times in AoC 2019
 * puzzles.
 */
#include "utils.hpp"
#include "intcode.hpp"

using namespace std;

/*------------------------------------------------------------------------------
  VARIABLES
------------------------------------------------------------------------------*/

int IntcodeProgram::INSTANCE_ID = 0;
const char* OPERATION_NAMES[] = { "", "add", "mult", "read", "write",
  "jump_if_true", "jump_if_false", "set_if_lt", "set_if_eq",
  "offset_relative_base" };

/*------------------------------------------------------------------------------
  FUNCTIONS
------------------------------------------------------------------------------*/
/* Constructor/Destructor ----------------------------------------------------*/
IntcodeProgram::IntcodeProgram(vector<long long> program, bool debug) {
  for (int i = 0; i < program.size(); i++) {
    this->program_[i] = program[i];
    this->initialProgram_[i] = program[i];
  }
  this->debug_ = debug;
  this->id_ = IntcodeProgram::INSTANCE_ID++;
  this->instructionPtr_ = 0;
  this->relativeBase_ = 0;
  this->isRunning_ = false;
}

IntcodeProgram::~IntcodeProgram() {}

/* Getters and Setters -------------------------------------------------------*/
long long IntcodeProgram::getProgramData(int index) const {
  map<int,long long>::const_iterator it = this->program_.find(index);
  return (it != this->program_.end()) ? it->second : 0;
}
int IntcodeProgram::getInstructionPtr() const {
  return this->instructionPtr_;
}
std::vector<long long> IntcodeProgram::getOutput() const {
  return this->output_;
}
long long IntcodeProgram::getOutputAt(int index) const {
  if (index < 0) index += this->output_.size();
  return (this->output_.size() == 0) ? -1 : this->output_[index];
}
long long IntcodeProgram::getLastOutput() const {
  return (this->output_.size() == 0) ? -1 : this->output_.back();
}
long long IntcodeProgram::popMemory() {
  long long v = this->memory_.front();
  // this->memory_.pop_front();
  this->memory_.erase(this->memory_.begin());
  return v;
}
void IntcodeProgram::setProgramData(int index, long long value) {
  this->program_[index] = value;
}
void IntcodeProgram::pushMemory(long long value) {
  this->memory_.push_back(value);
}
void IntcodeProgram::pushMemory(std::vector<long long> values) {
  this->memory_.insert(this->memory_.end(), values.begin(), values.end());
}
void IntcodeProgram::insertMemory(long long value) {
  this->memory_.insert(this->memory_.begin(), value);
}
void IntcodeProgram::insertMemory(std::vector<long long> values) {
  this->memory_.insert(this->memory_.begin(), values.begin(), values.end());
}

/* Private methods -----------------------------------------------------------*/
bool IntcodeProgram::processOpcode_() {
  // get the current instruction
  int instruction = (int)(this->getProgramData(this->instructionPtr_));
  // extract the operation code (opcode) and check for halt or error
  int opcode = instruction % 10 + 10 * ((instruction / 10) % 10);
  if (opcode == OP_HALT) {
    if (this->debug_) {
      cout << "\n[ 99 ] - Exiting\n";
    }
    this->instructionPtr_ = -1;
    return false;
  }
  if (opcode < 1 || opcode > 9) {
    this->instructionPtr_ = -1;
    return false;
  }
  // get the information on this operation for further process and debug
  const char* opname = OPERATION_NAMES[opcode];
  this->mode1_ = (instruction / 100) % 10;
  this->mode2_ = (instruction / 1000) % 10;
  this->mode3_ = (instruction / 10000) % 10;
  this->inputId_ = 0;
  if (this->debug_) {
    this->debugStr_ = strFormat(
      "\n[ % 3d ] - inst = %05d :: op = %s (%d), modes = %d, %d, %d\n",
      this->instructionPtr_, instruction, opname, opcode, this->mode1_,
      this->mode2_, this->mode3_
    );
  }
  // prepare the pause mode as false (could be modified by some operations)
  bool pause = false;
  // execute the right operation depending on the opcode
  this->instructionPtr_++;
  long long va, vb, vc, vm;
  switch (opcode) {
    case OP_ADD:
      va = this->getValue_();
      vb = this->getValue_();
      vc = this->getValue_(true);
      this->setProgramData(vc, va + vb);
      break;
    case OP_MULT:
      va = this->getValue_();
      vb = this->getValue_();
      vc = this->getValue_(true);
      this->setProgramData(vc, va * vb);
      break;
    case OP_READ:
      vm = this->popMemory();
      if (this->memory_.size() == 0) {
        this->instructionPtr_ = -1;
        return false;
      }
      va = this->getValue_(true);
      this->setProgramData(va, vm);
      break;
    case OP_WRITE:
      va = this->getValue_();
      this->pushOutput_(va);
      pause = true;
      break;
    case OP_JUMP_IF_TRUE:
      va = this->getValue_();
      vb = this->getValue_();
      if (va != 0) {
        this->instructionPtr_ = vb;
      }
      break;
    case OP_JUMP_IF_FALSE:
      va = this->getValue_();
      vb = this->getValue_();
      if (va == 0) {
        this->instructionPtr_ = vb;
      }
      break;
    case OP_SET_IF_LT:
      va = this->getValue_();
      vb = this->getValue_();
      vc = this->getValue_(true);
      this->setProgramData(vc, (va < vb) ? 1 : 0);
      break;
    case OP_SET_IF_EQ:
      va = this->getValue_();
      vb = this->getValue_();
      vc = this->getValue_(true);
      this->setProgramData(vc, (va == vb) ? 1 : 0);
      break;
    case OP_OFFSET_RELATIVE_BASE:
      this->relativeBase_ += this->getValue_();
      break;
  }
  
  if (this->debug_) {
    cout << this->debugStr_ << '\n';
  }
  
  return pause;
}

void IntcodeProgram::getIndex_(long long& index, int& mode) {
  // extract the mode for this input (and check if there are no more inputs for
  // this instruction; if so: abort)
  switch (this->inputId_) {
    case 0: mode = this->mode1_; break;
    case 1: mode = this->mode2_; break;
    case 2: mode = this->mode3_; break;
    default: mode = -1; break;
  }
  if (mode == -1) {
    index = -1;
    return;
  }
  // process the index depending on the mode
  switch (mode) {
    case 0:
      index = this->getProgramData(this->instructionPtr_);
      break;
    case 1:
      index = (long long)(this->instructionPtr_);
      break;
    case 2:
      index = this->getProgramData(this->instructionPtr_) + (long long)(this->relativeBase_);
      break;
    default:
      break;
  }
  // increase pointers
  this->instructionPtr_++;
  this->inputId_++;
}

long long IntcodeProgram::getValue_(bool keepIndex) {
  // get the index and mode
  long long index;
  int mode;
  this->getIndex_(index, mode);
  if (index == -1) {
    return -1;
  }
  // if necessary, apply the index as an address in the program code
  long long val = (!keepIndex) ? this->getProgramData((int)(index)) : index;
  // (fill the debug string in case of debug mode)
  this->debugStr_ += strFormat(" arg%d=%ld (idx=%ld, mode=%d) ;", this->inputId_,
    val, index, mode);
  return val;
}

void IntcodeProgram::pushOutput_(long long value) {
  this->output_.push_back(value);
}

/* Public methods ------------------------------------------------------------*/
int IntcodeProgram::run(unsigned int pauseEvery) {
  if (pauseEvery == 0) {
    // process while operation is not "halt"
    while (this->instructionPtr_ != -1) {
      this->processOpcode_();
    }
  } else {
    int nPause = 0;
    bool pause;
    // process while operation is not "halt"
    while (this->instructionPtr_ != -1) {
      pause = this->processOpcode_();
      if (pause) {
        nPause++;
      }
      if (pauseEvery == nPause) {
        nPause = 0;
        return (this->instructionPtr_ != -1) ? 1 : 0;
      }
    }
  }
  return 0;
}

int IntcodeProgram::runMultiple(vector<IntcodeProgram*> instances) {
  int nextInstance;
  long long output;
  int nInstances = instances.size();
  // if we stopped just before halting, we simply terminate the program and go
  // to the next instance
  if (this->output_.size() > 0 && this->instructionPtr_ == -1) {
    nextInstance = (this->id_ + 1) % nInstances;
    output = this->getLastOutput();
    instances[nextInstance]->pushMemory(output);
    return nextInstance;
  }
  // else we continue running the program from where we stopped
  bool pause;
  while (this->instructionPtr_ != -1) {
    pause = this->processOpcode_();
    // if we reached the halt op for the last instance
    if (this->instructionPtr_ == -1 && this->id_ == nInstances - 1) {
      return -1;
    }
    // else if we need to temporary pause the execution of this instance
    if (pause || this->instructionPtr_ == -1) {
      nextInstance = (this->id_ + 1) % nInstances;
      output = this->getLastOutput();
      instances[nextInstance]->pushMemory(output);
      return nextInstance;
    }
  }
  return -1;
}

void IntcodeProgram::checkRunning(unsigned int phase) {
  if (!this->isRunning_) {
    this->pushMemory(phase);
    this->isRunning_ = true;
  }
}

void IntcodeProgram::reset() {
  this->program_.clear();
  for (auto i : this->initialProgram_) {
    this->program_[i.first] = i.second;
  }
  this->memory_.clear();
  this->output_.clear();
  this->instructionPtr_ = 0;
  this->relativeBase_ = 0;
  this->isRunning_ = false;
}

void IntcodeProgram::resetOutput() {
  this->output_.clear();
}

void IntcodeProgram::printProgram() const {
  cout << "Current pointer: " << this->instructionPtr_ << '\n';
  for (auto p : this->program_) {
    cout << "program[" << p.first << "] = " << p.second << '\n';
  }
}

void IntcodeProgram::printMemory() const {
  for (int i = 0; i < this->memory_.size(); i++) {
    cout << "memory[" << i << "] = " << this->memory_[i] << '\n';
  }
}

void IntcodeProgram::printOutput() const {
  for (int i = 0; i < this->output_.size(); i++) {
    cout << "output[" << i << "] = " << this->output_[i] << '\n';
  }
}

File: 2019/Javascript/utils/intcode.js
const _ = require('lodash')

let INSTANCE_ID = 0 // class variable that is common to all instances
let INPUT = null // specific value for the input instruction
let OUTPUTS = [] // outputs of the program
const OPERATIONS = {
  1: { name: 'add', op: (a, b) => a + b, n: 3 },
  2: { name: 'mult', op: (a, b) => a * b, n: 3 },
  3: { name: 'read', op: () => INPUT, n: 1 },
  4: { name: 'write', op: null, n: 1 },
  5: { name: 'jump_if_true', op: null, n: 2 },
  6: { name: 'jump_if_false', op: null, n: 2 },
  7: { name: 'set_if_lt', op: (a, b) => a < b, n: 3 },
  8: { name: 'set_if_eq', op: (a, b) => a === b, n: 3 },
  9: { name: 'offset_relative_base', op: null, n: 1 },
  99: { name: 'halt', op: null, n: null }
}

/**
 * Util class to represent a program instance with its own instructions,
   memory, run state and instruction pointer. Allows for multiple instances
   in parallel to interact without overwriting data.
 * @class
 */
class IntcodeProgram {
  
  /**
   * @constructs IntcodeProgram
   * @param {array(int)} program - Original Intcode program to execute (will
   *                               be copied to avoid in-place modification).
   * @param {bool} debug - Whether or not the IntcodeProgram should debug its
   *                       execution at each instruction processing.
   */
  constructor(program, debug=false) {
    this.id = INSTANCE_ID++
    this.program = _.reduce(program, (acc, inst, idx) => {
      return { ...acc, [idx]: inst }
    }, {})
    this.memory = []
    this.output = []
    this.modes = []
    this.instructionPtr = 0
    this.isRunning = false
    this.relativeBase = 0
    this.debug = debug
    
    this._initialProgram = _.cloneDeep(this.program)
    this._inputId = 0
    this._debugStr = ''
  }
  
  /**
   * Resets the program instance in case you want to re-run the same program
     with a fresh start.
   */
  reset() {
    this.instructionPtr = 0
    this.output = []
    this.memory = []
    this.isRunning = false
    this.relativeBase = 0
    this.program = _.cloneDeep(this._initialProgram)
  }
  
  /**
   * Inserts a value in the instance's memory, in first position.
   * @param {array(int)} program - Original Intcode program to execute (will
   *                               be copied to avoid in-place modification).
   */
  setProgram(program) {
    this._initialProgram = _.reduce(program, (acc, inst, idx) => {
      return { ...acc, [idx]: inst }
    }, {})
    this.reset()
  }
  
  /**
   * Resets the output of the program to a blank slate.
   */
  resetOutput() {
    this.output = []
  }
  
  /**
   * Pushes one or more value(s) in the instance's memory, at the end.
   * @param {int | array(int)} data - Value(s) to append.
   */
  pushMemory(data) {
    if (_.isArray(data)) {
      this.memory = _.concat(this.memory, data)
    } else {
      this.memory.push(data)
    }
  }
  
  /**
   * Inserts one or more value(s) in the instance's memory, in first position.
   * @param {int | array(int)} data - Value(s) to insert.
   */
  insertMemory(data) {
    if (_.isArray(data)) {
      this.memory = _.concat(data, this.memory)
    } else {
      this.memory.unshift(data)
    }
  }
  
  /**
   * Checks if the instance is already running or if it should be initialized
   * with its phase setting.
   * @param {int} phase - Phase setting for this instance.
   */
  checkRunning(phase) {
    if (!this.isRunning) {
      this.insertMemory(phase)
      this.isRunning = true
    }
  }
  
  /**
   * Gets a value in the instance's program at a given position (if the position
     is out of range, returns 0).
   * @param {int} index - Position to get.
   * @returns {int} - Program data value.
   */
  programGetData(index) {
    return _.get(this.program, index, 0)
  }
  
  /**
   * Sets a value in the instance's program at a given position.
   * @param {int} index - Position to get.
   * @param {int} data - Value to insert.
   */
  programSetData (index, data) {
    this.program[index] = data
  }
  
  /**
   * Runs the instance by executing its Intcode program from start to finish
     (until it halts).
   * @param {null | int} pauseEvery - If not None, number of output digits to
   *                                  store before pausing. If None, the
   *                                  execution should proceed until it reached
   *                                  the halt operation.
   */
  run(pauseEvery=null) {
    let nPause = 0, pause
    while (this.instructionPtr !== null) {
      pause = this.processOpcode()
      // check for pause
      if (pause) {
        nPause++
      }
      if (pauseEvery === nPause) {
        nPause = 0
        return (this.instructionPtr !== null) ? 'pause' : null
      }
      if (this.instructionPtr === -1) {
        return null
      }
    }
    return null
  }
  
  /**
   * Runs the instance by executing its Intcode program either from scratch or
     from where it last stopped, as part of a pool of instances that feed each
     other with output to input connection.
   * @param {array(IntcodeProgram)} instances - List of all program instances
   *                                             in the pool.
   * @returns {int} - Index of the next instance in the pool to run, if any.
   */
  runMultiple(instances) {
    // if we stopped just before halting, we simply terminate the program
    // and go to the next instance
    let nextInstance
    if (!_.isEmpty(this.output) && this.instructionPtr === null) {
      nextInstance = (this.id + 1) % instances.length
      instances[nextInstance].pushMemory(_.last(this.output))
      return nextInstance
    }
    // else we continue running the program from where we stopped
    let pause
    while (this.instructionPtr !== null) {
      pause = this.processOpcode()
      // if we reached the halt op for the last instance
      if (this.instructionPtr === null && this.id === instances.length - 1) {
        return -1
      }
      // else if we need to temporary pause the execution of this instance
      if (pause || this.instructionPtr === null) {
        nextInstance = (this.id + 1) % instances.length
        instances[nextInstance].pushMemory(_.last(this.output))
        return nextInstance
      }
      // else if we errored
      if (this.instructionPtr === -1) {
        return null
      }
    }
  }
  
  /**
   * Gets the index corresponding to the cell pointed by the current instruction
     pointer plus the current input (in "address" or "immediate value").
   * @returns {object(int, int)} - Index and mode of the next input.
   */
  getIndex() {
    // check if there are no more inputs for this instruction; if so: abort
    if (_.isEmpty(this.modes)) {
      return { idx: null, mode: null }
    }
    // extract the mode for this input
    const mode = _.head(this.modes)
    this.modes = _.slice(this.modes, 1)
    // process the index depending on the mode
    let idx
    switch (mode) {
      case 0:
        idx = this.programGetData(this.instructionPtr)
        break
      case 1:
        idx = this.instructionPtr
        break
      case 2:
        idx = this.programGetData(this.instructionPtr) + this.relativeBase
        break
      default:
        return { idx: null, mode: null }
    }
    // increase the current instruction pointer
    this.instructionPtr++
    // increase the input id (for debug)
    this._inputId++
    return { idx, mode }
  }
  
  /**
   * Gets the "address" or "immediate value" for a given set of inputs and data.
   * @param {bool} keepIndex - Whether or not the function should keep the index
   *                           as is, or interpret it as an address in the
   *                           program.
   * @returns {int} - Program data value.
   */
  getValue(keepIndex=false) {
    // get the index and mode
    const { idx, mode } = this.getIndex()
    if (idx === null) {
      return null
    }
    // if necessary, apply the index as an address in the program code
    const val = (keepIndex) ? idx : this.programGetData(idx)
    // (fill the debug string in case of debug mode)
    this._debugStr += ` arg${this._inputId}=${val} (idx=${idx}, mode=${mode}) ;`
    return val
  }
  
  /**
   * Processes the next instruction in the program with the current memory and
     instruction pointer.
   * @returns {bool} - Whether or not the program should pause (if pause is
   *                   activated).
   */
  processOpcode() {
    // get the current instruction
    const instruction = this.program[this.instructionPtr].toString()
    // extract the operation code (opcode) and check for halt or error
    const opcode = parseInt(instruction.substring(instruction.length - 2))
    if (!_.has(OPERATIONS, opcode)) {
      this.instructionPtr = -1
      return false
    }
    // get the information on this operation for further process
    const { name, op, n } = OPERATIONS[opcode]
    const modes = instruction.substring(0, instruction.length - 2)
    this.modes = _.concat(
      _.map(modes.split('').reverse().join(''), (m) => Number(m)),
      new Array(n - modes.length).fill(0)
    )
    // prepare the debug string in case the debug mode is active
    this._inputId = 0
    this._debugStr = (
      `[ ${this.instructionPtr.toString().padStart(3, ' ')} ]`
      + ` - inst = ${instruction.padStart(5, '0')} `
      + `:: op = ${name} (${opcode}), `
      + `modes = ${this.modes}\n`
    )
    // prepare the pause mode as False (could be modified by some operations)
    let pause = false
    // execute the right operation depending on the opcode
    this.instructionPtr++
    let a, b, c, m, v
    switch (opcode) {
      case 99:
        this.instructionPtr = null
        break
      case 1: // add
      case 2: // multiply
        a = this.getValue(); b = this.getValue(); c = this.getValue(true)
        this.programSetData(c, op(a, b))
        break
      case 3: // read
        if (_.isEmpty(this.memory)) {
          return false
        }
        a = this.getValue(true)
        m = _.head(this.memory)
        this.memory = _.slice(this.memory, 1)
        this.programSetData(a, m)
        break
      case 4: // print
        v = this.getValue()
        this.output.push(v)
        pause = true
        break
      case 5: // jump if true
        a = this.getValue(); b = this.getValue()
        if (a !== 0) {
          this.instructionPtr = b
        }
        break
      case 6: // jump if false
        a = this.getValue(); b = this.getValue()
        if (a === 0) {
          this.instructionPtr = b
        }
        break
      case 7: // set if less than
      case 8: // set if equal
        a = this.getValue(); b = this.getValue(); c = this.getValue(true)
        this.programSetData(c, (op(a, b)) ? 1 : 0)
        break
      case 9: // relative base offset
        this.relativeBase += this.getValue()
      default:
        return false
    }
    
    if (this.debug) {
      console.log(this._debugStr)
    }
    
    return pause
  }
  
}

/**
 * Resets the global ID for program instances.
 */
IntcodeProgram.resetID = () => {
  INSTANCE_ID = 0
}

module.exports = {
  IntcodeProgram
}
